forem/forem
# Have specs run without Redis and refactor ConsumerApp and Device

## What type of PR is this? (check all applicable)


## Description

The goal is to remove `Redis` as a dependency to run tests, with better isolation from third party servers.

This PR started as "Well, I'll install `fakeredis` and that's it" and it definitely didn't end there :-)

Long story short: the main "blocker" to run specs without Redis at all was `rpush-redis`. While trying to find a way to test with the gem but without the actual Redis server, I realized many related tests were actually testing `rpush` and not how we interacted with it.

Thus to remove the server as a hard dependency for specs, I decided to refactor the code and the tests and keep the boundaries as much as possible within our app and not trespass into the gem.

## QA Instructions, Screenshots, Recordings

Most is automatic, but you can start the website, go to http://localhost:3000/admin/consumer_apps and make sure everything works.

## Added tests?

- [x] Yes, many :D
      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
